presto-native-execution/presto_cpp/main/types/ExpressionOptimizer.cpp

```diff
@@ -0,0 +1,150 @@
+    "X-Presto-Expression-Optimizer-Level";
+
+static const velox::expression::MakeFailExpr kMakeFailExpr =
+    [](const std::string& error,
+       const velox::TypePtr& type) -> velox::core::TypedExprPtr {
+  return std::make_shared<velox::core::CastTypedExpr>(
+      type,
+      std::vector<velox::core::TypedExprPtr>{
+          std::make_shared<velox::core::CallTypedExpr>(
+              velox::UNKNOWN(),
+              std::vector<velox::core::TypedExprPtr>{
+                  std::make_shared<velox::core::ConstantTypedExpr>(
+                      velox::VARCHAR(), error)},
+              fmt::format(
+                  "{}fail",
+                  SystemConfig::instance()->prestoDefaultNamespacePrefix()))},
+      false);
+};
+
+// Tries to evaluate `expr`, irrespective of its determinism, to a constant
+// value.
+velox::VectorPtr tryEvaluateToConstant(
+    const velox::core::TypedExprPtr& expr,
+    velox::core::QueryCtx* queryCtx,
+    velox::memory::MemoryPool* pool) {
+  auto data =
+      velox::BaseVector::create<velox::RowVector>(velox::ROW({}), 1, pool);
+  velox::core::ExecCtx execCtx{pool, queryCtx};
+  velox::exec::ExprSet exprSet({expr}, &execCtx);
+  velox::exec::EvalCtx evalCtx(&execCtx, &exprSet, data.get());
+
+  const velox::SelectivityVector singleRow(1);
+  std::vector<velox::VectorPtr> results(1);
+  exprSet.eval(singleRow, evalCtx, results);
+  return results.at(0);
+}
+
+protocol::RowExpressionOptimizationResult getRowExpressionOptimizationResult(
+    const RowExpressionPtr& rowExpression,
+    const std::string& optimizerLevel,
+    const VeloxExprConverter& veloxExprConverter,
+    const expression::VeloxToPrestoExprConverter& veloxToPrestoExprConverter,
+    velox::core::QueryCtx* queryCtx,
+    velox::memory::MemoryPool* pool) {
+  protocol::RowExpressionOptimizationResult result;
+  const auto expr = veloxExprConverter.toVeloxExpr(rowExpression);
+  auto optimized =
+      velox::expression::optimize(expr, queryCtx, pool, kMakeFailExpr);
+
+  if (optimizerLevel == kEvaluated) {
+    try {
+      const auto evalResult = tryEvaluateToConstant(optimized, queryCtx, pool);
+      optimized = std::make_shared<velox::core::ConstantTypedExpr>(evalResult);
+    } catch (const velox::VeloxException& e) {
+      result.expressionFailureInfo =
+          toNativeSidecarFailureInfo(translateToPrestoException(e));
+      result.optimizedExpression = nullptr;
+      return result;
+    } catch (const std::exception& e) {
+      result.expressionFailureInfo =
+          toNativeSidecarFailureInfo(translateToPrestoException(e));
+      result.optimizedExpression = nullptr;
+      return result;
+    }
+  }
+
+  result.optimizedExpression =
+      veloxToPrestoExprConverter.getRowExpression(optimized, rowExpression);
+  return result;
+}
+
+} // namespace
+
+void optimizeExpressions(
+    const proxygen::HTTPHeaders& httpHeaders,
+    const json::array_t& inputRowExpressions,
+    proxygen::ResponseHandler* downstream,
```

Comment:
This method should have a better abstract signature. We should return the json::array_t result and set the downstream response in the caller.

On the input side, its better to extract the optimizerLevel and timezone in the caller and pass them as params to this function.